// x-client-transaction-id/script/errors.js
"use strict";
/**
 * Custom error types for x-client-transaction-id.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.XMigrationRedirectionError = exports.XMigrationFormError = exports.XHomePageFetchError = exports.SiteVerificationKeyNotFoundError = exports.OnDemandFileUrlResolutionError = exports.OnDemandFileFetchError = exports.KeyByteIndicesExtractionError = exports.InterpolationInputError = exports.IndicesNotInitializedError = exports.HandleXMigrationError = exports.ClientTransactionNotInitializedError = exports.ClientTransactionInitializationError = exports.ClientTransactionError = exports.AnimationFrameDataError = void 0;
class ClientTransactionError extends Error {
    constructor(message, options = {}) {
        super(message, options.cause ? { cause: options.cause } : undefined);
        Object.defineProperty(this, "code", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.name = new.target.name;
        this.code = options.code ?? "CLIENT_TRANSACTION_ERROR";
    }
}
exports.ClientTransactionError = ClientTransactionError;
class ClientTransactionInitializationError extends ClientTransactionError {
    constructor(message, options = {}) {
        super(message, {
            code: options.code ?? "CLIENT_TRANSACTION_INITIALIZATION_ERROR",
            cause: options.cause,
        });
    }
}
exports.ClientTransactionInitializationError = ClientTransactionInitializationError;
class OnDemandFileUrlResolutionError extends ClientTransactionInitializationError {
    constructor() {
        super("Unable to resolve the X ondemand chunk URL from the homepage runtime.", { code: "ONDEMAND_FILE_URL_RESOLUTION_ERROR" });
    }
}
exports.OnDemandFileUrlResolutionError = OnDemandFileUrlResolutionError;
class OnDemandFileFetchError extends ClientTransactionInitializationError {
    constructor(url, status, statusText) {
        super(`Unable to fetch the X ondemand chunk from "${url}": ${status} ${statusText}.`, { code: "ONDEMAND_FILE_FETCH_ERROR" });
        Object.defineProperty(this, "url", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "status", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "statusText", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.url = url;
        this.status = status;
        this.statusText = statusText;
    }
}
exports.OnDemandFileFetchError = OnDemandFileFetchError;
class KeyByteIndicesExtractionError extends ClientTransactionInitializationError {
    constructor() {
        super("Unable to extract key byte indices from the X ondemand chunk.", { code: "KEY_BYTE_INDICES_EXTRACTION_ERROR" });
    }
}
exports.KeyByteIndicesExtractionError = KeyByteIndicesExtractionError;
class SiteVerificationKeyNotFoundError extends ClientTransactionInitializationError {
    constructor() {
        super("Unable to find the twitter-site-verification meta tag in the homepage document.", { code: "SITE_VERIFICATION_KEY_NOT_FOUND_ERROR" });
    }
}
exports.SiteVerificationKeyNotFoundError = SiteVerificationKeyNotFoundError;
class IndicesNotInitializedError extends ClientTransactionError {
    constructor() {
        super("ClientTransaction indices are not initialized. Call initialize() before generating animation data.", { code: "INDICES_NOT_INITIALIZED_ERROR" });
    }
}
exports.IndicesNotInitializedError = IndicesNotInitializedError;
class AnimationFrameDataError extends ClientTransactionInitializationError {
    constructor(rowIndex) {
        super(`Unable to build animation data for row ${rowIndex}. The homepage animation markup may have changed.`, { code: "ANIMATION_FRAME_DATA_ERROR" });
        Object.defineProperty(this, "rowIndex", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.rowIndex = rowIndex;
    }
}
exports.AnimationFrameDataError = AnimationFrameDataError;
class ClientTransactionNotInitializedError extends ClientTransactionError {
    constructor() {
        super("ClientTransaction has not been initialized. Call initialize() or use ClientTransaction.create() first.", { code: "CLIENT_TRANSACTION_NOT_INITIALIZED_ERROR" });
    }
}
exports.ClientTransactionNotInitializedError = ClientTransactionNotInitializedError;
class HandleXMigrationError extends ClientTransactionError {
    constructor(message, options = {}) {
        super(message, {
            code: options.code ?? "HANDLE_X_MIGRATION_ERROR",
            cause: options.cause,
        });
    }
}
exports.HandleXMigrationError = HandleXMigrationError;
class XHomePageFetchError extends HandleXMigrationError {
    constructor(status, statusText) {
        super(`Unable to fetch the X homepage: ${status} ${statusText}.`, { code: "X_HOMEPAGE_FETCH_ERROR" });
        Object.defineProperty(this, "status", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "statusText", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.status = status;
        this.statusText = statusText;
    }
}
exports.XHomePageFetchError = XHomePageFetchError;
class XMigrationRedirectionError extends HandleXMigrationError {
    constructor(status, statusText) {
        super(`Unable to follow the X migration redirect: ${status} ${statusText}.`, { code: "X_MIGRATION_REDIRECTION_ERROR" });
        Object.defineProperty(this, "status", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "statusText", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.status = status;
        this.statusText = statusText;
    }
}
exports.XMigrationRedirectionError = XMigrationRedirectionError;
class XMigrationFormError extends HandleXMigrationError {
    constructor(status, statusText) {
        super(`Unable to submit the X migration form: ${status} ${statusText}.`, { code: "X_MIGRATION_FORM_ERROR" });
        Object.defineProperty(this, "status", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "statusText", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.status = status;
        this.statusText = statusText;
    }
}
exports.XMigrationFormError = XMigrationFormError;
class InterpolationInputError extends ClientTransactionError {
    constructor(fromLength, toLength) {
        super(`Interpolation requires arrays of the same length, but received ${fromLength} and ${toLength}.`, { code: "INTERPOLATION_INPUT_ERROR" });
        Object.defineProperty(this, "fromLength", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "toLength", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.fromLength = fromLength;
        this.toLength = toLength;
    }
}
exports.InterpolationInputError = InterpolationInputError;
